File: packages/ai_config/catalog/normalize.py
from typing import Dict


def normalize_catalog(catalog: Dict) -> Dict:
    provider_map = {}
    for provider in catalog.get("providers", []):
        normalized_id = provider["id"].strip().lower()
        provider_map[normalized_id] = {**provider, "id": normalized_id}

    defaults = catalog.get("defaults", {})
    default_provider = defaults.get("provider", "").strip().lower()

    model_map = {}
    for model in catalog.get("models", []):
        provider_id = (model.get("provider_id") or model.get("providerId") or default_provider).strip().lower()
        model_id = model["id"].strip().lower()
        key = f"{provider_id}/{model_id}"
        model_map[key] = {**model, "provider_id": provider_id, "id": model_id}

    return {
        **catalog,
        "providers": sorted(provider_map.values(), key=lambda p: p["id"]),
        "models": sorted(model_map.values(), key=lambda m: (m.get("provider_id", ""), m["id"])),
    }

File: packages/ai-config/src/catalog/normalize.ts
import type { Catalog } from '../domain/catalog.js';

export function normalizeCatalog(catalog: Catalog): Catalog {
  const providerById = new Map<string, (typeof catalog.providers)[number]>();
  for (const provider of catalog.providers) {
    const normalizedId = provider.id.trim().toLowerCase();
    providerById.set(normalizedId, { ...provider, id: normalizedId });
  }

  const modelByKey = new Map<string, (typeof catalog.models)[number]>();
  for (const model of catalog.models) {
    const providerId = (model.providerId ?? catalog.defaults?.provider ?? '').trim().toLowerCase();
    const normalizedModelId = model.id.trim().toLowerCase();
    const key = `${providerId}/${normalizedModelId}`;
    modelByKey.set(key, {
      ...model,
      id: normalizedModelId,
      providerId,
    });
  }

  return {
    ...catalog,
    providers: [...providerById.values()].sort((a, b) => a.id.localeCompare(b.id)),
    models: [...modelByKey.values()].sort((a, b) => {
      const byProvider = (a.providerId ?? '').localeCompare(b.providerId ?? '');
      return byProvider !== 0 ? byProvider : a.id.localeCompare(b.id);
    }),
  };
}
